src/libp2p_transport_tcp.erl

```diff
@@ -11,28 +16,204 @@
         ]).
 
 -record(tcp_state, {
+          addr_info :: {string(), string()},
           socket :: gen_tcp:socket(),
           transport :: atom()
          }).
 
--type state() :: #tcp_state{}.
+-type tcp_state() :: #tcp_state{}.
+
+-record(state, {
+          tid :: ets:tab(),
+          stun_sup ::supervisor:sup_ref(),
+          stun_txns=#{} :: maps:map(),
+          observed_addrs=sets:new() :: sets:set()
+         }).
+
 
 -define(CONFIG_SECTION, tcp).
 
--spec start_listener(pid(), string(), ets:tab()) -> {ok, [string()], pid()} | {error, term()}.
-start_listener(Sup, Addr, TID) ->
+%% API
+%%
+
+-spec new_connection(inet:socket()) -> libp2p_connection:connection().
+new_connection(Socket) ->
+    {ok, LocalAddr} = inet:sockname(Socket),
+    {ok, RemoteAddr} = inet:peername(Socket),
+    libp2p_connection:new(?MODULE, #tcp_state{addr_info={to_multiaddr(LocalAddr), to_multiaddr(RemoteAddr)},
+                                              socket=Socket,
+                                              transport=ranch_tcp}).
+
+-spec start_listener(pid(), string()) -> {ok, [string()], pid()} | {error, term()}.
+start_listener(Pid, Addr) ->
+    gen_server:call(Pid, {start_listener, Addr}).
+
+-spec connect(pid(), string(), [libp2p_swarm:connect_opt()], pos_integer()) -> {ok, libp2p_session:pid()} | {error, term()}.
+connect(Pid, MAddr, Options, Timeout) ->
+    gen_server:call(Pid, {connect, MAddr, Options, Timeout}, infinity).
+
+
+-spec match_addr(string()) -> {ok, string()} | false.
+match_addr(Addr) when is_list(Addr) ->
+    match_protocols(multiaddr:protocols(multiaddr:new(Addr))).
+
+match_protocols([A={_, _}, B={"tcp", _} | _]) ->
+    {ok, multiaddr:to_string([A, B])};
+match_protocols(_) ->
+    false.
+
+
+%% libp2p_connection
+%%
+-spec send(tcp_state(), iodata(), non_neg_integer()) -> ok | {error, term()}.
+send(#tcp_state{socket=Socket, transport=Transport}, Data, Timeout) ->
+    Transport:setopts(Socket, [{send_timeout, Timeout}]),
+    Transport:send(Socket, Data).
+
+-spec recv(tcp_state(), non_neg_integer(), pos_integer()) -> {ok, binary()} | {error, term()}.
+recv(#tcp_state{socket=Socket, transport=Transport}, Length, Timeout) ->
+    Transport:recv(Socket, Length, Timeout).
+
+-spec close(tcp_state()) -> ok.
+close(#tcp_state{socket=Socket, transport=Transport}) ->
+    Transport:close(Socket).
+
+-spec close_state(tcp_state()) -> open | closed.
+close_state(#tcp_state{socket=Socket}) ->
+    case inet:peername(Socket) of
```

Comment:
Any reason you use peername here?

Reply:
It was the easiest way to find out if the socket was still open or closed